cmd/pinniped-proxy/src/cache.rs

```diff
@@ -0,0 +1,134 @@
+// Copyright 2020-2022 the Kubeapps contributors.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{collections::HashMap, hash::Hash, sync::RwLock};
+
+/// An unlimited cache that can be RwLocked across threads.
+///
+/// Importantly, checking the cache does not require a write-lock
+/// (unlike the [`Cached` trait's `cache_get`](https://github.com/jaemk/cached/blob/f5911dc3fbc03e1db9f87192eb854fac2ee6ac98/src/lib.rs#L203))
+#[derive(Default)]
+struct LockableCache<K, V>(RwLock<HashMap<K, V>>);
```

Comment:
This struct is a [NewType pattern](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-the-newtype-pattern-to-implement-external-traits-on-external-types) which is a zero-cost abstraction (ie. no penalty at run-time) that defines a new type as a thin wrapper - a 1-tuple of another type, so that we can add our caching functions (`get`, `insert`) to this type without needing an extra struct. So in this case, a LockableCache is really just a read-write lock wrapping a hash, but one which behaves like a cache.